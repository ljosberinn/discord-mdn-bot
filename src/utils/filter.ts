/**
 * same as Array#filter but works on all iterables, returns an iterable
 *
 * @export
 * @template T input iterable
 * @template U output iterable
 * @param {(item: T, index: number, iter: Iterable<T>) => any} fn
 * @param {Iterable<T>} iter
 */
export function* filterʹ<T>(
  fn: (item: T, index: number, iter: Iterable<T>) => T,
  iter: Iterable<T>
) {
  let i = 0;
  for (const item of iter) {
    if (fn(item, i++, iter)) yield item;
  }
}

export function filter<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U
) {
  return (iter: Iterable<T>) => filterʹ(fn, iter);
}
