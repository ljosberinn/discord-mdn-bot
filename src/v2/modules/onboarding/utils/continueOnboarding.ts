import type { GuildMember, Guild, TextChannel } from 'discord.js';

import { ONBOARDING_CHANNEL } from '../../../env';
import type { UserStateType } from '../db/user_state';
import { handleIntroduction } from '../steps/handleIntroduction';
import { handleOnboarded } from '../steps/handleOnboarded';
import { handleRoleSelection } from '../steps/handleRoleSelection';
import { handleStart } from '../steps/handleStart';

export async function continueOnboarding(
  guild: Guild,
  member: GuildMember,
  oldState: UserStateType,
  fromStart = false
): Promise<void> {
  const channel = guild.channels.resolve(ONBOARDING_CHANNEL) as TextChannel;

  switch (oldState.state) {
    case 'START':
      handleStart(guild, member, oldState, fromStart);
      return;
    case 'INTRODUCTION':
      handleIntroduction(guild, member, oldState, fromStart);
      return;
    case 'ROLE_SELECTION':
      handleRoleSelection(guild, member, oldState, fromStart);
      return;
    case 'ONBOARDED':
      handleOnboarded(guild, member, oldState, fromStart);
      return;
    default:
      console.error("Shouldn't have gotten here D:");
  }
}
