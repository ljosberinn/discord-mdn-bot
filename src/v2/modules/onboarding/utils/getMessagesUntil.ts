import type { Message, TextChannel } from 'discord.js';
import { filter } from 'domyno';

export async function* getMessagesUntil(
  channel: TextChannel,
  date: Date,
  limit = 100
): AsyncIterableIterator<Message> {
  let before;
  const timestamp = date.getTime();

  const msgAfter = filter<Message>(
    (msg: Message) => msg.createdTimestamp > timestamp
  );

  while (true) {
    // this is fine this is a async generator function
    // eslint-disable-next-line no-await-in-loop
    const messages = await channel.messages.fetch({ limit, before });
    if (messages.size === 0) {
      return;
    }
    yield* msgAfter(messages.values());
    before = messages.last().id;
  }
}
