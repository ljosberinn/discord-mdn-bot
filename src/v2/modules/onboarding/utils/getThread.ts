import type { Guild, TextChannel, ThreadChannel } from 'discord.js';

import { ONBOARDING_CHANNEL } from '../../../env';

export async function getThread(
  guild: Guild,
  id: string
): Promise<ThreadChannel> {
  const channel = (await guild.channels.fetch(
    ONBOARDING_CHANNEL
  )) as TextChannel;

  return channel.threads.fetch(id);
}
