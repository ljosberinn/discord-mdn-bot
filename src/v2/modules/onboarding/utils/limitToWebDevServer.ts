import { SERVER_ID } from '../../../env';

export function limitToWebDevServer<
  HasGuildId extends { guild: { id: string } },
  Output,
  VArgs extends ReadonlyArray<unknown>
>(
  fn: (g: HasGuildId, ...vargs: VArgs) => Output
): (a: HasGuildId, ...vargs: VArgs) => Output {
  return (x: HasGuildId, ...args: VArgs) => {
    if (x.guild.id !== SERVER_ID) {
      return;
    }

    return fn(x, ...args);
  };
}
