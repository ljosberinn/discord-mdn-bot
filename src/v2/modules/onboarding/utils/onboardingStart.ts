import { get, upsert } from '../../../cache';
import type { GenericCacheType } from '../../../cache/model';
import { SERVER_ID } from '../../../env';

type OnboardingStartCache = {
  meta: {
    onboardingStart: number;
  };
} & GenericCacheType;

const _getOnboardingCache = (): Promise<OnboardingStartCache> =>
  get({
    guild: SERVER_ID,
    type: 'ONBOARDING_START',
    user: '',
  }) as unknown as Promise<OnboardingStartCache>;

export const getOnboardingStart = async (): Promise<number | undefined> => {
  const cache = await _getOnboardingCache();

  return cache?.meta.onboardingStart;
};

export const setOnboardingStart = (): Promise<OnboardingStartCache> =>
  upsert({
    expiresAt: Number.MAX_SAFE_INTEGER,
    guild: SERVER_ID,
    type: 'ONBOARDING_START',
    user: '',
    meta: {
      onboardingStart: Date.now(),
    },
  }) as unknown as Promise<OnboardingStartCache>;
