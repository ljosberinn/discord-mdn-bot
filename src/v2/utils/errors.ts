import { REPO_LINK } from '../env.js';

export const invalidResponse =
  'sorry, your request could not be processed. Please try again at a later time.';
export const noResults = (search: string): string =>
  `sorry, could not find anything for \`${search}\`.`;
export const unknownError = `sorry, something went wrong. If this issue persists, please file an issue at ${REPO_LINK}`;
export const missingRightsDeletion =
  'insufficient permissions: unable to delete message.';
export const userNotFound =
  'sorry, your user ID cannot be found within the database.';
