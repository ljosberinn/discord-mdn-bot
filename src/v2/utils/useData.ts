import { addBreadcrumb } from '@sentry/node';
import type {
  HeadersInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';
import fetch from 'node-fetch';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env.js';
import { Cache } from './Cache.js';

const apiCache = new Cache({
  checkperiod: Number.parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  maxKeys: Number.parseInt(API_CACHE_ENTRIES_LIMIT, 10),
  stdTTL: Number.parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper):FetchWithFormat<TParsedResponse> => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get(casedCacheKey);

  if (cachedResponse) {
    return (async () => cachedResponse) as FetchWithFormat<TParsedResponse>;
  }

  return async (url, fetchOptions) => {
    addBreadcrumb({
      category: 'query',
      data: typeof url === 'string' ? { url } : undefined,
      level: 'info',
      timestamp: Date.now(),
    });

    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    // eslint-disable-next-line no-console
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    // eslint-disable-next-line no-console
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse as TParsedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = <T>(
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    } as unknown as Promise<UnknownData<T>>;
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    }as unknown as Promise<UnknownData<T>>;
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    }as unknown as Promise<UnknownData<T>>;
  }
};

const useData = <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeadersInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};

export default useData;
