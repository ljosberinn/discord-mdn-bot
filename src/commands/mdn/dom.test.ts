import * as DomParser from 'dom-parser';

import * as errors from '../../utils/errors';
import { getSearchUrl } from '../../utils/urlTools';
import useData from '../../utils/useData';

import { queryBuilder } from './dom';
import { getChosenResult } from '../../utils/discordTools';

jest.mock('dom-parser');
jest.mock('../../utils/urlTools');
jest.mock('../../utils/useData');

const mockGetSearchUrl: jest.MockedFunction<typeof getSearchUrl> = getSearchUrl as any;
const mockUseData: jest.MockedFunction<typeof useData> = useData as any;
const mockChoose: jest.MockedFunction<typeof getChosenResult> = getChosenResult as any;

describe('handleMDNQuery', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  beforeEach(() => {
    mockGetSearchUrl.mockReturnValue('Search Term');
  });

  afterEach(() => jest.resetAllMocks());

  test('replies with invalid response error if search URL fails', async () => {
    mockUseData.mockResolvedValue({
      error: true,
      json: null,
      text: null,
    });

    await queryBuilder()(msg, 'Search Term');

    expect(msg.reply).toHaveBeenCalledWith(errors.invalidResponse);
    expect(msg.channel.send).not.toHaveBeenCalled();
  });

  test('replies with 0 documents found', async () => {
    mockUseData.mockResolvedValue({
      error: false,
      json: null,
      text: 'Example',
    });

    await queryBuilder(text => {
      expect(text).toEqual('Example');
      return {
        isEmpty: true,
        meta: '',
        results: [],
      };
    })(msg, 'Search Term');

    expect(msg.reply).toBeCalledWith(errors.noResults('Search Term'));
  });

  test('responds with list embedded', async () => {
    mockUseData.mockResolvedValue({
      error: false,
      json: null,
      text: 'Example',
    });

    await queryBuilder(
      text => {
        expect(text).toEqual('Example');
        return {
          isEmpty: false,
          meta: '',
          results: [
            {
              getElementsByClassName(
                className: string
              ): DomParser.Node[] | null {
                if (className === 'result-title') {
                  return [
                    {
                      getAttribute() {
                        return 'http://example.com';
                      },
                      textContent: '',
                    } as any,
                  ];
                }
                return [
                  {
                    textContent: 'Some markdown',
                  } as any,
                ];
              },
            } as any,
          ],
        };
      },
      () => ({
        excerpt: '',
        title: 'Example',
        url: 'http://www.example.com',
      }),
      () =>
        [
          {
            url: 'http://www.example.com',
          },
        ] as any
    )(msg, 'Search Term');

    expect(msg.channel.send).toHaveBeenCalledTimes(1);
    const sentMessage = msg.channel.send.mock.calls[0][0];
    expect(sentMessage).toMatchSnapshot();
  });
});
