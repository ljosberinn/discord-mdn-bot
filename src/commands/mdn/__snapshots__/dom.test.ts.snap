// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleMDNQuery responds with list embedded 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 8638706,
    "description": "1. [**Example** - ](http://www.example.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`‚ùå\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://avatars0.githubusercontent.com/u/7565578",
      "text": "",
    },
    "title": undefined,
    "url": "Search Term",
  },
}
`;
